webapp/api/github.py

```diff
@@ -0,0 +1,139 @@
+import flask
+from urllib.parse import urlencode
+from webapp import api
+from requests.exceptions import RequestException
+
+
+class GithubAPIError(RequestException):
+    pass
+
+
+class GithubAPI:
+    """
+    Provides authentication for GitHub users. Helper methods are also provided
+    for checking organization access and getting user data from the Github API.
+    """
+
+    BASE_URL = "https://api.github.com/"
+    AUTH_URL = "https://github.com/login/oauth/"
+
+    def __init__(
+        self, client_id, client_secret, session=api.requests.Session()
+    ):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.session = session
+        self.session.headers["Accept"] = "application/json"
+
+    def authorize(self, scope=None, state=None):
+        """
+        Redirect to Github and request access to a user's data.
+        """
+        params = {
+            "client_id": self.client_id,
+        }
+
+        if scope:
+            params["scope"] = scope
+
+        if state:
+            params["state"] = state
+
+        auth_url = f"{self.AUTH_URL}authorize?{urlencode(params)}"
+
+        return flask.redirect(auth_url)
+
+    def verify_auth(self, code):
+        """
+        Handles response after the redirect to Github. This response
+        determines if the user has allowed this application access. If we
+        were then we send a POST request for the access_key used to
+        authenticate requests to Github.
+        """
+        data = {
+            "code": code,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+
+        response = self.request("POST", self.AUTH_URL, "access_token", data)
+
+        response.raise_for_status()
+
+        data = response.json()
+
+        if "error" in data:
+            raise GithubAPIError(data["error_description"], response=response)
+
+        return data["access_token"]
+
+    def request(
+        self, method, base_url, resource, data={}, params={}, access_token=None
+    ):
+        """
+        Makes a raw HTTP request and returns the response.
+        """
+        url = base_url + resource
+
+        if access_token is None:
+            access_token = flask.session.get("github_auth_secret")
```

Comment:
I think for this class to remain neatly encapsulated, it would be best to avoid any Flask-level functions if we can. I think what should happen here is that getting the secret from the session should be something that happens more at the view level and then you should pass it through as an argument wherever you're instantiating the `GithubAPI` class.

Reply:
@nottrobin maybe this is related with your previous comment but right now if I add a property `self.access_token` that also can be set with the constructor does that could cause the users to have the same access_token? because we are instantiating the class only once?